File: locaura-backend/src/Consumer/models/Consumer.model.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IConsumerAddress {
    _id?: mongoose.Types.ObjectId;
    label: string; // e.g. "Home", "Office"
    line1: string;
    line2?: string;
    city: string;
    state: string;
    pincode: string;
    is_default: boolean;
    location?: {
        type: 'Point';
        coordinates: [number, number]; // [longitude, latitude]
    };
}

export interface ICartItem {
    variant_id: mongoose.Types.ObjectId;
    quantity: number;
    product_name?: string;
    brand_name?: string;
    price?: number;
    thumb_url?: string;
}

export interface IConsumerCart {
    store_id?: mongoose.Types.ObjectId;
    store_name?: string;
    items: ICartItem[];
    subtotal?: number;
    platform_fee?: number;
    delivery_fee?: number;
    total?: number;
}

export interface IConsumer extends Document {
    _id: mongoose.Types.ObjectId;
    phone: string;
    otp?: string;
    otp_expiry?: Date;
    phone_verified: boolean;
    
    consumer_name?: string;
    email?: string;
    
    addresses: IConsumerAddress[];
    cart: IConsumerCart;
    
    fcm_token?: string;
    
    status: 'active' | 'suspended' | 'deleted';
    
    createdAt: Date;
    updatedAt: Date;
}

const AddressSchema = new Schema<IConsumerAddress>(
    {
        label: { type: String, required: true, trim: true },
        line1: { type: String, required: true },
        line2: { type: String },
        city: { type: String, required: true },
        state: { type: String, required: true },
        pincode: { type: String, required: true },
        is_default: { type: Boolean, default: false },
        location: {
            type: { type: String, enum: ['Point'], default: 'Point' },
            coordinates: { type: [Number], required: false } // [longitude, latitude]
        }
    },
    { _id: true }
);

const CartItemSchema = new Schema<ICartItem>(
    {
        variant_id: { type: Schema.Types.ObjectId, ref: 'ChildProduct', required: true },
        quantity: { type: Number, required: true, min: 1 },
        product_name: { type: String },
        brand_name: { type: String },
        price: { type: Number },
        thumb_url: { type: String }
    },
    { _id: false }
);

const CartSchema = new Schema<IConsumerCart>(
    {
        store_id: { type: Schema.Types.ObjectId, ref: 'Store' },
        store_name: { type: String },
        items: { type: [CartItemSchema], default: () => [] },
        subtotal: { type: Number, default: 0 },
        platform_fee: { type: Number, default: 0 },
        delivery_fee: { type: Number, default: 0 },
        total: { type: Number, default: 0 }
    },
    { _id: false }
);

const ConsumerSchema = new Schema<IConsumer>(
    {
        phone: {
            type: String,
            required: true,
            unique: true,
            trim: true,
            match: [/^\d{10}$/, 'Please fill a valid 10-digit mobile number']
        },
        otp: { type: String },
        otp_expiry: { type: Date },
        phone_verified: { type: Boolean, default: false },

        consumer_name: { type: String, trim: true },
        email: { 
            type: String, 
            trim: true, 
            lowercase: true,
            match: [/^\S+@\S+\.\S+$/, 'Please use a valid email address'] 
        },

        addresses: { type: [AddressSchema], default: () => [] },
        cart: { type: CartSchema, default: () => ({ items: [] }) },

        fcm_token: { type: String },

        status: { type: String, enum: ['active', 'suspended', 'deleted'], default: 'active' },
    },
    { timestamps: true }
);

// Search and performance indexes
ConsumerSchema.index({ phone: 1 });
ConsumerSchema.index({ email: 1 });
ConsumerSchema.index({ status: 1 });

export const Consumer = mongoose.model<IConsumer>('Consumer', ConsumerSchema);

File: locaura_parter/lib/Consumer/features/auth/data/models/consumer.model.dart
// ignore_for_file: invalid_annotation_target

import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:locaura_parter/Consumer/features/auth/domain/entities/consumer.entity.dart';

part 'consumer.model.freezed.dart';
part 'consumer.model.g.dart';

@freezed
class ConsumerModel with _$ConsumerModel {
  const factory ConsumerModel({
    @JsonKey(name: '_id') required String id,
    required String phone,
    @JsonKey(name: 'phone_verified') required bool phoneVerified,
    @JsonKey(name: 'consumer_name') String? name,
    String? email,
    @JsonKey(name: 'avatar_url') String? avatarUrl,
    @Default([]) List<AddressModel> addresses,
    CartModel? cart,
  }) = _ConsumerModel;

  factory ConsumerModel.fromJson(Map<String, dynamic> json) =>
      _$ConsumerModelFromJson(json);
}

extension ConsumerModelX on ConsumerModel {
  ConsumerEntity toEntity({required String token}) => ConsumerEntity(
        id: id,
        phone: phone,
        phoneVerified: phoneVerified,
        name: name,
        email: email,
        avatarUrl: avatarUrl,
        token: token,
        addresses: addresses.map((a) => a.toEntity()).toList(),
        cart: cart?.toEntity(),
      );
}



@freezed
class ConsumerProfileResponse with _$ConsumerProfileResponse {
  const factory ConsumerProfileResponse({
    String? token,
    required ConsumerModel consumer,
  }) = _ConsumerProfileResponse;

  factory ConsumerProfileResponse.fromJson(Map<String, dynamic> json) =>
      _$ConsumerProfileResponseFromJson(json);
}

extension ConsumerProfileResponseX on ConsumerProfileResponse {
  ConsumerEntity toEntity() => consumer.toEntity(token: token!);
}

@freezed
class AddressModel with _$AddressModel {
  const factory AddressModel({
    @JsonKey(name: '_id') String? id,
    required String label,
    required String line1,
    String? line2,
    required String city,
    required String state,
    required String pincode,
    @JsonKey(name: 'is_default') required bool isDefault,
    LocationModel? location,
  }) = _AddressModel;

  factory AddressModel.fromJson(Map<String, dynamic> json) =>
      _$AddressModelFromJson(json);
}

extension AddressModelX on AddressModel {
  AddressEntity toEntity() => AddressEntity(
        id: id,
        label: label,
        line1: line1,
        line2: line2,
        city: city,
        state: state,
        pincode: pincode,
        isDefault: isDefault,
        location: location?.toEntity(),
      );
}
  
extension LocationModelX on LocationModel {
  LocationEntity toEntity() => LocationEntity(
        type: type,
        coordinates: coordinates,
      );
}


@freezed
class LocationModel with _$LocationModel {
  const factory LocationModel({
    @Default('Point') String type,
    required List<double> coordinates,
  }) = _LocationModel;

  factory LocationModel.fromJson(Map<String, dynamic> json) =>
      _$LocationModelFromJson(json);
}

@freezed
class CartModel with _$CartModel {
  const factory CartModel({
    @JsonKey(name: 'store_id') String? storeId,
    @JsonKey(name: 'store_name') String? storeName,
    @Default([]) List<CartItemModel> items,
    @Default(0.0) double subtotal,
    @Default(0.0) double total,
    @Default(0.0) double delivery_fee,
    @Default(0.0) double platform_fee,
  }) = _CartModel;

  factory CartModel.fromJson(Map<String, dynamic> json) =>
      _$CartModelFromJson(json);
}

extension CartModelX on CartModel {
  ConsumerCartEntity toEntity() => ConsumerCartEntity(
        storeId: storeId,
        storeName: storeName,
        items: items.map((i) => i.toEntity()).toList(),
        subtotal: subtotal.toDouble(),
        total: total.toDouble(),
        delivery_fee: delivery_fee.toDouble(),
        platform_fee: platform_fee.toDouble(),
      );
}

@freezed
class CartItemModel with _$CartItemModel {
  const factory CartItemModel({
    @JsonKey(name: 'variant_id') required String variantId,
    required int quantity,
    @JsonKey(name: 'product_id') String? productId,
    @JsonKey(name: 'product_name') String? productName,
    @JsonKey(name: 'brand_name') String? brandName,
    String? size,
    String? color,
    double? price,
    @JsonKey(name: 'original_price') double? originalPrice,
    @JsonKey(name: 'thumb_url') String? thumbUrl,
    @JsonKey(name: 'variant_sku') String? variantSku,
    @JsonKey(name: 'variant_label') String? variantLabel,
    @JsonKey(name: 'total_price') double? totalPrice,
    @JsonKey(name: 'stock_available') int? stockAvailable,
  }) = _CartItemModel;

  factory CartItemModel.fromJson(Map<String, dynamic> json) =>
      _$CartItemModelFromJson(json);
}

extension CartItemModelX on CartItemModel {
  ConsumerCartItemEntity toEntity() => ConsumerCartItemEntity(
        variantId: variantId,
        quantity: quantity,
        productId: productId??'',
        productName: productName ?? 'Unknown Product',
        brandName: brandName ?? '',
        size: size ?? '',
        color: color ?? '',
        price: price?.toDouble() ?? 0.0,
        originalPrice: originalPrice?.toDouble(),
        thumbUrl: thumbUrl,
        variantSku: variantSku,
        variantLabel: variantLabel,
        totalPrice: totalPrice?.toDouble() ?? ((price?.toDouble() ?? 0.0) * quantity),
        stockAvailable: stockAvailable,
      );
}
